core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/recipe/RecipeExecutionFailureHandler.java

```diff
@@ -0,0 +1,117 @@
+package com.sequenceiq.cloudbreak.service.cluster.flow.recipe;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.stereotype.Component;
+
+import com.google.common.collect.Sets;
+import com.sequenceiq.cloudbreak.domain.HostGroup;
+import com.sequenceiq.cloudbreak.domain.InstanceGroup;
+import com.sequenceiq.cloudbreak.domain.InstanceMetaData;
+import com.sequenceiq.cloudbreak.domain.Recipe;
+import com.sequenceiq.cloudbreak.orchestrator.exception.CloudbreakOrchestratorException;
+import com.sequenceiq.cloudbreak.orchestrator.model.RecipeModel;
+
+@Component
+public class RecipeExecutionFailureHandler {
+
+    public Set<RecipeExecutionFailure> handleError(CloudbreakOrchestratorException exception,
+            Map<HostGroup, List<RecipeModel>> hostgroupToRecipeMap, Set<InstanceGroup> instanceGroups) {
+
+        Set<RecipeExecutionFailure> result = Sets.newHashSet();
+
+        exception.getNodesWithErrors().asMap().forEach((node, errors) -> {
+            errors.forEach(error -> {
+                Optional<String> installPhase = getInstallPhase(error);
+                if (installPhase.isPresent()) {
+                    String phase = installPhase.get();
+                    String recipeName = StringUtils.substringBetween(error, "sh -x /opt/scripts/" + phase + "/", " 2>&1 |");
+                    Set<String> possibleFailingHostgroups = getPossibleFailingHostgroupsByRecipeName(hostgroupToRecipeMap, recipeName);
+                    possibleFailingHostgroups.stream()
+                            .map(hostGroupName -> {
+                                Optional<Recipe> recipe = getRecipeOfHostgroupByRecipeName(hostgroupToRecipeMap, recipeName, hostGroupName);
+                                Optional<InstanceMetaData> instanceMetaData = getInstanceMetaDataByHost(instanceGroups, node, hostGroupName);
+                                if (recipe.isPresent() && instanceMetaData.isPresent()) {
+                                    return new RecipeExecutionFailure(recipe.get(), instanceMetaData.get());
+                                }
+                                return null;
+                            })
+                            .filter(Objects::nonNull)
+                            .forEach(result::add);
+                }
+            });
+        });
+
+        return result;
```

Comment:
My implementation looks like this:
```
    public Set<RecipeExecutionFailure> handleError(CloudbreakOrchestratorException exception,
            Map<HostGroup, List<RecipeModel>> hostgroupToRecipeMap, Set<InstanceGroup> instanceGroups) {
        return exception.getNodesWithErrors().asMap().entrySet().stream().map(nodeWithErrors -> {
            Map<String, Optional<String>> errorsWithPhase = nodeWithErrors.getValue().stream().collect(Collectors.toMap(e -> e, this::getInstallPhase));
            return errorsWithPhase.entrySet().stream()
                    .filter(e -> e.getValue().isPresent())
                    .map(e -> StringUtils.substringBetween(e.getKey(), "sh -x /opt/scripts/" + e.getValue().get() + '/', " 2>&1 |"))
                    .collect(Collectors.toMap(n -> n, n -> getPossibleFailingHostgroupsByRecipeName(hostgroupToRecipeMap, n))).entrySet().stream()
                    .map(e -> e.getValue().stream()
                        .map(hg -> {
                            Optional<Recipe> recipe = getRecipeOfHostgroupByRecipeName(hostgroupToRecipeMap, e.getKey(), hg);
                            Optional<InstanceMetaData> instanceMetaData = getInstanceMetaDataByHost(instanceGroups, nodeWithErrors.getKey(), hg);
                            if (recipe.isPresent() && instanceMetaData.isPresent()) {
                                return new RecipeExecutionFailure(recipe.get(), instanceMetaData.get());
                            }
                            return null;
                        })
                        .filter(Objects::nonNull)
                        .collect(Collectors.toSet())
                    )
                    .flatMap(Collection::stream)
                    .collect(Collectors.toSet());
        })
        .flatMap(Collection::stream)
        .collect(Collectors.toSet());
    }
```
Theres is no if statement inside, and `forEach` at all